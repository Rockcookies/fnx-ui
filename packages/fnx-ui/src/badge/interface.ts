import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
