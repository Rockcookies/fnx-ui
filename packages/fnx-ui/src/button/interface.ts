import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
