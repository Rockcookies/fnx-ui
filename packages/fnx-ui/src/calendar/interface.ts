import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
