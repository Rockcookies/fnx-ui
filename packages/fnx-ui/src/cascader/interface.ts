import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
