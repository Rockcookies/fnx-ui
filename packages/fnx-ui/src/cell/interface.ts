import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
