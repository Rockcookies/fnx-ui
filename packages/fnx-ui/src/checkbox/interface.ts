import { HTMLAttributes, ReactNode } from 'react';
import InternalCheckboxGroup from './CheckboxGroup';

export type CheckboxValue = string | number | boolean;

export interface CheckboxBaseProps {
	disabled?: boolean;
	direction?: 'vertical' | 'horizontal';
	checkedIcon?: ReactNode;
	icon?: ReactNode;
	iconSize?: string | number;
	iconPosition?: 'left' | 'right';
	iconShape?: 'square' | 'round' | 'plain';
	iconCheckedColor?: string;
	labelDisabled?: boolean;
}

export interface CheckboxComponentProps extends CheckboxBaseProps {
	value?: CheckboxValue;
	checked?: boolean;
	defaultChecked?: boolean;
	onChange?: (checked: boolean) => void;
	skipGroup?: boolean;
}

export type CheckboxProps = CheckboxComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'value'>;

export interface CheckboxRef {
	element: HTMLDivElement | null;
	toggle: () => void;
}

export interface CheckboxGroupComponentProps<
	T extends CheckboxValue = CheckboxValue,
> extends CheckboxBaseProps {
	value?: T[];
	defaultValue?: T[];
	onChange?: (v: T[]) => void;
	maxCheckedCount?: number;
}

export type CheckboxGroupProps<T extends CheckboxValue = CheckboxValue> =
	CheckboxGroupComponentProps<T> &
		Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'>;

export interface CheckboxGroupContextData extends CheckboxBaseProps {
	value?: CheckboxValue[];
	registerValue?: (value: CheckboxValue) => void;
	cancelValue?: (value: CheckboxValue) => void;
	toggleValue?: (isCheck: boolean, value: CheckboxValue) => void;
}
export interface CheckboxGroupRef {
	element: HTMLDivElement | null;
	toggleAll: (options?: { checkAll?: boolean }) => void;
}

export interface CheckboxExportExtra {
	Group: typeof InternalCheckboxGroup;
}
