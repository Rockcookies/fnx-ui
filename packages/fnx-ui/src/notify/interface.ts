import { HTMLAttributes } from 'react';
import { PopupComponentProps } from '../popup';

export type NotifyType = 'primary' | 'success' | 'warning' | 'danger';

export interface NotifyComponentProps extends PopupComponentProps {
	type?: NotifyType;
	message?: string;
	duration?: number;
	color?: string;
	background?: string;
}

export type NotifyProps = NotifyComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'type'>;

export interface NotifyContext {
	show: (props: string | NotifyProps) => NotifyInstance;
}

export interface NotifyExportExtra {
	show: (props: string | NotifyProps) => NotifyInstance;
	useNotify: () => NotifyContext;
	allowMultiple: (allow: boolean) => void;
	clearAll: () => void;
}

export interface NotifyInstance {
	update: (props: NotifyProps) => void;
	clear: () => void;
}
