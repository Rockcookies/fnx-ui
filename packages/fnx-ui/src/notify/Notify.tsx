import React, { useEffect, useState } from 'react';
import Popup from '../popup';
import { classnames, createBEM } from '../utils/namespace';
import { createDefaultsForwardRef } from '../utils/react';
import { NotifyProps } from './interface';

const NS = 'fnx-notify';
const bem = createBEM(NS);

const Notify = createDefaultsForwardRef<
	HTMLDivElement,
	NotifyProps,
	Required<Pick<NotifyProps, 'type' | 'visible' | 'duration' | 'lockScroll'>>
>(
	'Notify',
	{
		type: 'danger',
		visible: false,
		duration: 2000,
		lockScroll: false,
	},
	(
		{
			type,
			visible,
			duration,
			lockScroll,
			// optionals
			mountTo,
			message,
			color,
			background,
			children,
			className,
			style,
			...restProps
		},
		ref,
	) => {
		const [showing, setShowing] = useState(false);

		useEffect(() => {
			setShowing(visible);

			if (visible && duration > 0) {
				const timer = setTimeout(() => {
					setShowing(false);
				}, duration);

				return () => {
					clearTimeout(timer);
				};
			}
		}, [visible, duration]);

		return (
			<Popup
				visible={showing}
				mountTo={mountTo}
				className={classnames(bem([type]), className)}
				renderOnShow={true}
				destroyOnHide={true}
				overlay={false}
				position="top"
				lockScroll={lockScroll}
				style={{
					color,
					background,
					...style,
				}}
				{...restProps}
				ref={ref}
			>
				{message || children}
			</Popup>
		);
	},
);

export default Notify;
