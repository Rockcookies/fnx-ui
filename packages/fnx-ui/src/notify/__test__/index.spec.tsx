import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React, { FC } from 'react';
import TestsDOM from '../../../test/dom';
import ConfigProvider from '../../config-provider';
import Notify from '../index';
import { NotifyContext, NotifyInstance } from '../interface';

describe('<Notify />', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	function renderNotifyList(
		triggers: Array<{
			id: string;
			onClick: (ctx: NotifyContext) => void;
		}>,
	) {
		const Demo: FC = () => {
			const ctx = Notify.useNotify();

			return (
				<>
					{triggers.map(({ onClick, id }, idx) => (
						<div
							key={idx}
							id={id}
							onClick={() => {
								onClick(ctx);
							}}
						></div>
					))}
				</>
			);
		};

		const element = document.createElement('div');

		const { container } = render(
			<ConfigProvider mountTo={() => element}>
				<Demo />
			</ConfigProvider>,
		);

		return {
			element,
			fire: (id: string) => {
				const trigger = TestsDOM.mustQuerySelector(container, `#${id}`);

				act(() => {
					fireEvent.click(trigger);
				});
			},
		};
	}

	it('should render Notify correctly', () => {
		const { container, getByTestId } = render(
			<Notify data-testid="notify" visible />,
		);

		expect(getByTestId('notify')).toMatchSnapshot();

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('different display styles for different type', () => {
		const { container, getByTestId } = render(
			<Notify data-testid="notify" visible type="danger" />,
		);

		expect(getByTestId('notify').className).toContain('fnx-notify--danger');

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('color props', () => {
		const { container, getByTestId } = render(
			<Notify data-testid="notify" visible color="#ad0000" />,
		);

		expect(getByTestId('notify')).toMatchSnapshot();

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('onBeforeHide event is triggered correctly', () => {
		const onBeforeHide = jest.fn();

		const { getByTestId } = render(
			<Notify
				visible
				data-testid="notify"
				duration={1000}
				onBeforeHide={onBeforeHide}
			>
				Content
			</Notify>,
		);
		expect(getByTestId('notify')).toMatchSnapshot();

		act(() => {
			jest.advanceTimersByTime(1000);
		});

		expect(onBeforeHide).toBeCalledTimes(1);
	});

	it('should change default duration after calling update method', async () => {
		let _notify: NotifyInstance | undefined;

		const { element, fire } = renderNotifyList([
			{
				id: 'demo-notify',
				onClick: (ctx) => {
					_notify = ctx.show({
						id: 'notify',
						duration: 5000,
						message: 'Message',
					});
				},
			},
		]);

		fire('demo-notify');

		if (!_notify) {
			throw new Error('notify instance undefined');
		}

		const notify = _notify;

		expect(
			TestsDOM.mustQuerySelector(element, '#notify').innerHTML,
		).toEqual('Message');

		await waitFor(() =>
			notify.update({
				message: 'updateMessage',
			}),
		);

		expect(
			TestsDOM.mustQuerySelector(element, '#notify').innerHTML,
		).toEqual('updateMessage');
	});

	it('useToast', () => {
		const { element, fire } = renderNotifyList([
			{
				id: 'demo-notify',
				onClick: (ctx) => {
					ctx.show('success');
				},
			},
		]);

		fire('demo-notify');

		expect(element).toMatchSnapshot();
	});

	it('clear multiple notify', async () => {
		Notify.allowMultiple(true);

		const { element, fire } = renderNotifyList([
			{
				id: 'demo-notify',
				onClick: (ctx) => {
					ctx.show({ message: '1', duration: 0 });
					ctx.show({ message: '2', duration: 0 });
				},
			},
		]);

		fire('demo-notify');

		expect(element.querySelectorAll('.fnx-notify').length).toBe(2);

		await waitFor(() => Notify.clearAll());

		Notify.allowMultiple(false);
	});
});
