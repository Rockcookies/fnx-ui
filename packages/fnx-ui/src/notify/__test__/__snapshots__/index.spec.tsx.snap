// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notify /> color props 1`] = `
<div
  class="fnx-popup fnx-popup--top fnx-notify fnx-notify--danger fnx-popup--slide-top-enter fnx-popup--slide-top-enter-active"
  data-testid="notify"
  role="dialog"
  style="color: rgb(173, 0, 0); z-index: 2003;"
  tabindex="0"
/>
`;

exports[`<Notify /> color props 2`] = `<div />`;

exports[`<Notify /> different display styles for different type 1`] = `<div />`;

exports[`<Notify /> onBeforeHide event is triggered correctly 1`] = `
<div
  class="fnx-popup fnx-popup--top fnx-notify fnx-notify--danger fnx-popup--slide-top-enter fnx-popup--slide-top-enter-active"
  data-testid="notify"
  role="dialog"
  style="z-index: 2004;"
  tabindex="0"
>
  Content
</div>
`;

exports[`<Notify /> should render Notify correctly 1`] = `
<div
  class="fnx-popup fnx-popup--top fnx-notify fnx-notify--danger fnx-popup--slide-top-enter fnx-popup--slide-top-enter-active"
  data-testid="notify"
  role="dialog"
  style="z-index: 2001;"
  tabindex="0"
/>
`;

exports[`<Notify /> should render Notify correctly 2`] = `<div />`;

exports[`<Notify /> useToast 1`] = `
<div>
  <div>
    <div
      class="fnx-popup fnx-popup--top fnx-notify fnx-notify--danger fnx-popup--slide-top-enter fnx-popup--slide-top-enter-active"
      role="dialog"
      style="z-index: 2006;"
      tabindex="0"
    >
      success
    </div>
  </div>
</div>
`;
